### PR TITLE
Listen mode

### PR DESCRIPTION
**Pull request by magik6k**
Just a quick hack, probably can be improved in some ways, but it works for me

### CODE REVIEW

File: main.go [hunk 1]
```diff
@@ -38,12 +41,15 @@ var BootstrapAddresses = []string{
 var Info = log.New(ioutil.Discard, "I: ", 0)
 var Error = log.New(os.Stderr, "E: ", 0)
 
+const bufSize = 1 << 20
```
Comment (magik6k) [inline main.go:42]: Need this for some reason, makes transfer about 100x faster (~200mbit vs 2)

File: main.go [hunk 2]
```diff
@@ -151,7 +220,7 @@ func bootstrap(ctx context.Context, node host.Host) error {
 			done <- node.Connect(bctx, *pi)
 		}()
 	}
-	for _ = range BootstrapAddresses {
+	for range BootstrapAddresses {
```
Comment (Stebalien) [inline main.go:218]: What? I never new this was valid! Awesome!

File: main.go [hunk 3]
```diff
@@ -38,12 +41,15 @@ var BootstrapAddresses = []string{
 var Info = log.New(ioutil.Discard, "I: ", 0)
 var Error = log.New(os.Stderr, "E: ", 0)
 
+const bufSize = 1 << 20
+
 func main() {
 	var flags flag.FlagSet
 	flags.Usage = func() {
-		fmt.Fprintln(flags.Output(), "Usage: p2pcat [-v] [-routed] MULTIADDR PROTOCOL")
+		fmt.Fprintln(flags.Output(), "Usage: p2pcat [-v] [-routed] -listen/MULTIADDR PROTOCOL")
```
Comment (Stebalien) [inline main.go]: Nit:
```suggestion
		fmt.Fprintln(flags.Output(), "Usage: p2pcat [-v] [-routed] MULTIADDR PROTOCOL")
		fmt.Fprintln(flags.Output(), "  or:  p2pcat [-v] [-routed] -listen PROTOCOL")
```

(at least that's what some GNU commands do in cases like this)

File: main.go [hunk 4]
```diff
@@ -64,9 +70,68 @@ func main() {
...
+		err = bootstrap(ctx, node)
+		if err != nil {
+			return err
+		}
+		mhid, err := multihash.Cast([]byte(node.ID()))
+		if err != nil {
+			return err
+		}
+
+		go dhtclient.Provide(ctx, cid.NewCidV1(cid.Raw, mhid), true)
```
Comment (Stebalien) [inline main.go]: Shouldn't we just call `dht.Bootstrap(ctx)`?

File: main.go [hunk 5]
```diff
@@ -64,9 +70,68 @@ func main() {
...
+
+	node, err := libp2p.New(ctx)
+	if err != nil {
+		return err
+	}
+	defer func() {
+		node.Close()
+	}()
+	if routed {
+		dhtclient, err := dht.New(ctx, node, dhtopt.Client(false))
```
Comment (Stebalien) [inline main.go]: We should still be a DHT client. We don't really want to *answer* DHT queries.

**Review by Stebalien — CHANGES_REQUESTED** [review]
Nice!

**Comment by magik6k** [comment]
(done)

**Review by Stebalien — APPROVED** [review]

**Comment by Stebalien** [comment]
Awesome! Thanks!

**Pull request merged**